libraries/Microsoft.Bot.Builder/TurnContextServiceCollection.cs

```diff
@@ -65,7 +44,7 @@ public TurnContextServiceCollection()
             }
             catch(ArgumentException)
```

Comment:
This code is making my head hurt. 

Why would we change the semantics of Dictionary? if "Add()" on a dictionary throws an ArgumentException on duplicate Key ([which is does](https://msdn.microsoft.com/en-US/library/k7z0zy8k(v=vs.110).aspx)), why would we want to catch that, and throw a LESS useful exception with a hard-to-deal-with exception string? 

If we're on a "no custom exceptions" mission, then we should at the least, not introduce weird customer behavior. 

In **this** case, the behavior should be "don't catch" and just let the exception from the dictionary bubble up.

Reply:
Yes absolutely agree, I thought the same when i saw this code.

If you want to pull on that thread a little then you can ask about the value of the whole class (actually two classes and an interface.) That is pretty much where I was at.

Reply:
@cleemullins While I agree that would be better than what this PR does now, I still think that would also be a mistake. What does an exception with a stack trace bubbling up from a `Dictionary<,>.Add` call mean to me as a consumer of your library? Zip, now I gotta look at the stack trace and see what methods I was calling and decipher something logical from that. The library's job is to enrich the context of the exception with some domain specific details which, in this case, means speaking in more specific terms about the service. Even if you wanted to argue for just using `ArgumentException` here, which I could understand, the exception should still be thrown from where it is now, the exception message specialized and the dictionary's exception could be the inner exception if you really wanted to propagate that.

If the exception is just left to bubble up, it's technically leaking the internal implementation details of the library. Exceptions are part of the library's API contract with the outside world. What if the implementation was changed? Now the new implementation somehow has to craft an `ArgumentException` that looks just like the `Dictionary<,>`'s did for the rest of its life. It happens to "work" right now because our parameter, which would come through the `ArgumentException::ParamName` property is also called "key", but that's just a coincidence.

Nevermind that the .NET languages themselves are groomed towards handling strongly typed exceptions. What now... this pattern?

```CSharp
catch(Exception ex)
{
    if(ex.Message.Contains("something"))
    {
        // handle this
    }
    elseif(ex.Message.Contains("somethingelse"))
    {
        // handle that
    }
    else
   {
      throw;
   }
}
```

Even with modern exception filters now you still have this mess...

```CSharp
catch(Exception ex) when ex.Message.Contains("something")
{
    // handle this
}
catch(Exception ex) when ex.Message.Contains("somethingelse")
{
    // handle that
}
```

Finally, modern logging/monitoring solutions (e.g. Application Insights) intrinsically group by and provide rich dashboard experiences for .NET exceptions by their type? That's valuable, no? If everything is just `Exception` with a message that value is completely lost.

Just to close this out, I'm not necessarily arguing about _this specific exception scenario_. Could it be `ArgumentException` with a clear, domain specific message? Sure. I'm just deeply concerned about the sentiment that strongly typed exceptions are "useless" and therefore don't ever belong in the code base and want to arrive at some clarity on whether this is considered to be a blanket policy going forward.